### DISCUSSION TITLE
Can't get async iterator to work

### DISCUSSION THREAD

**Discussion by aspcompiler** [PyO3/pyo3#3352] [Questions] [answered]
### Bug Description

I am running 0.17.1. I have a very simple async iterator:

```
#[pyclass]
struct PyAsyncIter {
    count: usize,
}

#[pymethods]
impl PyAsyncIter {
    #[new]
    pub fn new() -> Self {
        PyAsyncIter { count: 0 }
    }

    fn __aiter__(slf: PyRef<'_, Self>) -> PyRef<'_, Self> {
        slf
    }

    fn __anext__(&mut self) -> Option<usize> {
        if self.count < 5 {
            self.count += 1;
            Some(self.count)
        } else {
            None
        }
    }
}

#[pymodule]
fn pyo3_test(_py: Python, m: &PyModule) -> PyResult<()> {
    m.add_class::<PyAsyncIter>()?;
    Ok(())
}
```

I got "TypeError: 'async for' received an invalid object from __anext__: int" when I run the following Python program:

```
from pyo3_test import PyAsyncIter
import asyncio

async def task():
  async for i in PyAsyncIter():
    print(i)

asyncio.run(task())
```

### Steps to Reproduce

See the Rust and Python programs in description

### Backtrace

```shell
TypeError: object int can't be used in 'await' expression

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/Cellar/python@3.10/3.10.12/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/Cellar/python@3.10/3.10.12/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "<stdin>", line 2, in task
TypeError: 'async for' received an invalid object from __anext__: int
```

### Your operating system and version

macOS Darwin 22.5.0

### Your Python version (`python --version`)

Python 3.10.12

### Your Rust version (`rustc --version`)

1.70.0

### Your PyO3 version

0.17.1

### How did you install python? Did you use a virtualenv?

brew

Yes, I used virtualenv.

**Comment by davidhewitt** [comment, marked as answer]
`__anext__` should return a python awaitable. Here you have implemented a sync iterable, consider switching to use `__iter__` and `__next__`.

**Reply by aspcompiler** [reply]
Thanks for the hint. I actually intentionally made it an async to that I can add async logic later. The follow worked. Added some boilerplate code and hope they could be abstracted away in future versions.

```
#[pyclass]
struct PyAsyncIter {
    count: Arc<Mutex<usize>>,
}

#[pymethods]
impl PyAsyncIter {
    #[new]
    pub fn new() -> Self {
        PyAsyncIter { count: Arc::new(Mutex::new(0)) }
    }

    fn __aiter__(slf: PyRef<'_, Self>) -> PyRef<'_, Self> {
        slf
    }

    fn __anext__(slf: PyRefMut<'_, Self>) -> PyResult<Option<PyObject>> {
        let count = slf.count.clone();
        let fut = pyo3_asyncio::tokio::future_into_py(slf.py(), async move {
            let mut count = count.lock().unwrap();
            if *count < 5 {
                *count += 1;
                Ok(Python::with_gil(|py| count.into_py(py)))
            } else {
                Err(PyStopAsyncIteration::new_err("stream exhausted"))
            }    
        })?;
        Ok(Some(fut.into()))
    }
}
```

**Reply by ShaneMurphy2** [reply]
What I didn't get was that `__anext__` needs to return an option. If you forget that the error message is particularly incoherent.

**Reply by davidhewitt** [reply]
@ShaneMurphy2 it's worth pointing out that PyO3 0.21 will not require an option. However, we are busy finishing off #3684 before 0.21 will be ready for release.

**Reply by ShaneMurphy2** [reply]
@davidhewitt Hm, any interesting reason that the Rust signature had to be different than the Python one in 0.20?

**Reply by davidhewitt** [reply]
Historical errors in design. We learn! 😂 

For some reading on the discussion where we eventually got it right:

https://github.com/PyO3/pyo3/issues/3190
https://github.com/PyO3/pyo3/pull/3661
https://pyo3.rs/main/migration#iteranextoutput-are-deprecated

**Reply by ShaneMurphy2** [reply]
Saw some of that, neat stuff. Love the crate, excited for 0.21!
